From 598c79ced6f99767005dbca937c5f0cd44b180d6 Mon Sep 17 00:00:00 2001
From: Tim Donohue <tim.donohue@lyrasis.org>
Date: Thu, 6 Aug 2026 15:42:19 -0500
Subject: [PATCH 1/4] Protect against path traversal attacks in "import" script
 when run from REST API or CLI. Validate the filename input is in an expected
 directory

(cherry picked from commit 285850b30d58acb5818f487f00f1d5f762f1ed84)
---
 .../org/dspace/app/itemimport/ItemImport.java | 14 ++++++++--
 .../dspace/app/itemimport/ItemImportCLI.java  | 28 +++++++++++++++----
 2 files changed, 34 insertions(+), 8 deletions(-)

diff --git a/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImport.java b/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImport.java
index 3824b5960a6a..7a188ded3c9e 100644
--- a/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImport.java
+++ b/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImport.java
@@ -13,6 +13,7 @@
 import java.io.InputStream;
 import java.net.URL;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.sql.SQLException;
 import java.time.Instant;
 import java.time.format.DateTimeFormatter;
@@ -40,6 +41,7 @@
 import org.dspace.handle.factory.HandleServiceFactory;
 import org.dspace.handle.service.HandleService;
 import org.dspace.scripts.DSpaceRunnable;
+import org.dspace.storage.secure.SecureFileAccess;
 import org.dspace.utils.DSpace;
 
 /**
@@ -352,8 +354,16 @@ protected void readZip(Context context, ItemImportService itemImportService) thr
                     validateZip(validationFileStream.get());
                 }
 
-                workFile = new File(itemImportService.getTempWorkDir() + File.separator
-                        + zipfilename + "-" + context.getCurrentUser().getID());
+                String workDir = itemImportService.getTempWorkDir();
+
+                // zipfilename is user controlled (via -z or -u param). So, we must validate the expected
+                // file path using SecureFileAccess to protect against path traversal attacks.
+                String fileName = zipfilename + "-" + context.getCurrentUser().getID();
+                String fileAbsolutePath = SecureFileAccess.calculateAbsolutePathUsingBaseDir(fileName, workDir);
+                Path validatedFilePath = SecureFileAccess.validatePathForWrite(fileAbsolutePath, List.of(workDir),
+                                                                               "ItemImport zip validation");
+
+                workFile = validatedFilePath.toFile();
                 FileUtils.copyInputStreamToFile(optionalFileStream.get(), workFile);
             } else {
                 throw new IllegalArgumentException(
diff --git a/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLI.java b/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLI.java
index bd29aa97fe48..4693a9e2aa26 100644
--- a/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLI.java
+++ b/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLI.java
@@ -11,6 +11,7 @@
 import java.io.FileInputStream;
 import java.io.InputStream;
 import java.net.URL;
+import java.nio.file.Path;
 import java.sql.SQLException;
 import java.util.List;
 import java.util.Optional;
@@ -23,6 +24,7 @@
 import org.dspace.content.Collection;
 import org.dspace.core.Context;
 import org.dspace.eperson.EPerson;
+import org.dspace.storage.secure.SecureFileAccess;
 
 /**
  * CLI variant for the {@link ItemImport} class.
@@ -103,8 +105,14 @@ protected void readZip(Context context, ItemImportService itemImportService) thr
         // If this is a zip archive, unzip it first
         if (zip) {
             if (!remoteUrl) {
+                // zipfilename is user controlled (via -z or -u param). So, we must validate the expected
+                // file path using SecureFileAccess to protect against path traversal attacks.
+                String fileAbsolutePath = SecureFileAccess.calculateAbsolutePathUsingBaseDir(zipfilename, sourcedir);
+                Path validatedZipPath = SecureFileAccess.validatePathForRead(fileAbsolutePath, List.of(sourcedir),
+                                                                             "ItemImportCLI zip validation");
+                File myZipFile = validatedZipPath.toFile();
+
                 // confirm zip file exists
-                File myZipFile = new File(sourcedir + File.separator + zipfilename);
                 if ((!myZipFile.exists()) || (!myZipFile.isFile())) {
                     throw new IllegalArgumentException(
                         "Error reading file, the file couldn't be found for filename: " + zipfilename);
@@ -120,8 +128,7 @@ protected void readZip(Context context, ItemImportService itemImportService) thr
 
                 workDir = new File(itemImportService.getTempWorkDir() + File.separator + TEMP_DIR
                         + File.separator + context.getCurrentUser().getID());
-                sourcedir = itemImportService.unzip(
-                        new File(sourcedir + File.separator + zipfilename), workDir.getAbsolutePath());
+                sourcedir = itemImportService.unzip(myZipFile, workDir.getAbsolutePath());
             } else {
                 // manage zip via remote url
                 Optional<InputStream> optionalFileStream = Optional.ofNullable(new URL(zipfilename).openStream());
@@ -134,10 +141,19 @@ protected void readZip(Context context, ItemImportService itemImportService) thr
                             validateZip(validationFileStream.get());
                         }
 
-                        workFile = new File(itemImportService.getTempWorkDir() + File.separator
-                                + zipfilename + "-" + context.getCurrentUser().getID());
+                        // zipfilename is user controlled (via -z or -u param). So, we must validate the expected
+                        // file path using SecureFileAccess to protect against path traversal attacks.
+                        String tempWorkDir = itemImportService.getTempWorkDir();
+                        String zipName = zipfilename + "-" + context.getCurrentUser().getID();
+                        String fileAbsolutePath = SecureFileAccess.calculateAbsolutePathUsingBaseDir(zipName,
+                                                                                                     tempWorkDir);
+                        Path validatedZipPath = SecureFileAccess
+                            .validatePathForWrite(fileAbsolutePath, List.of(tempWorkDir),
+                                                  "ItemImportCLI remote zip validation");
+
+                        workFile = validatedZipPath.toFile();
                         FileUtils.copyInputStreamToFile(optionalFileStream.get(), workFile);
-                        workDir = new File(itemImportService.getTempWorkDir() + File.separator + TEMP_DIR
+                        workDir = new File(tempWorkDir + File.separator + TEMP_DIR
                                 + File.separator + context.getCurrentUser().getID());
                         sourcedir = itemImportService.unzip(workFile, workDir.getAbsolutePath());
                     } else {

From eaa3350178bed85b7d54a002132c478772df065f Mon Sep 17 00:00:00 2001
From: Tim Donohue <tim.donohue@lyrasis.org>
Date: Thu, 6 Aug 2026 15:43:47 -0500
Subject: [PATCH 2/4] For consistency, update ItemImportServiceImpl.unzip() to
 use SecureFileAccess to validate unzip location. (NOTE: current code is
 already secure. This just updates the code to consistently use
 SecureFileAccess)

(cherry picked from commit 6b71fdfb7d5e234cb28be8f1e8b7b0cc58b31eb3)
---
 .../app/itemimport/ItemImportServiceImpl.java | 34 +++++++++++++++----
 1 file changed, 28 insertions(+), 6 deletions(-)

diff --git a/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java b/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java
index 5c1a0a4e5786..d156f30f9a48 100644
--- a/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java
+++ b/dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java
@@ -110,6 +110,7 @@
 import org.dspace.handle.service.HandleService;
 import org.dspace.scripts.handler.DSpaceRunnableHandler;
 import org.dspace.services.ConfigurationService;
+import org.dspace.storage.secure.SecureFileAccess;
 import org.dspace.workflow.WorkflowItem;
 import org.dspace.workflow.WorkflowService;
 import org.springframework.beans.factory.InitializingBean;
@@ -201,7 +202,7 @@ public void afterPropertiesSet() throws Exception {
         //Ensure tempWorkDir exists
         File tempWorkDirFile = new File(tempWorkDir);
         if (!tempWorkDirFile.exists()) {
-            boolean success = tempWorkDirFile.mkdir();
+            boolean success = tempWorkDirFile.mkdirs();
             if (success) {
                 logInfo("Created org.dspace.app.batchitemimport.work.dir of: " + tempWorkDir);
             } else {
@@ -1983,9 +1984,14 @@ public String unzip(File zipfile, String destDir) throws IOException {
             logError("Unable to create temporary directory: " + tempdir.getAbsolutePath());
         }
         String sourcedir = destinationDir + System.getProperty("file.separator") + zipfile.getName();
-        String zipDir = destinationDir + System.getProperty("file.separator") + zipfile.getName() + System
-            .getProperty("file.separator");
+        String zipDir = sourcedir + System.getProperty("file.separator");
+        File sourcedirFile = new File(sourcedir);
 
+        // Create the source directory we will be unzipping into. We must pre-create this directory to validate
+        // the final path of each zip entry using SecureFileAccess (see below)
+        if (!sourcedirFile.exists() && !sourcedirFile.mkdirs()) {
+            logError("Unable to create directory for unzipping: " + sourcedir);
+        }
 
         // 3
         String sourceDirForZip = sourcedir;
@@ -1996,13 +2002,22 @@ public String unzip(File zipfile, String destDir) throws IOException {
             while (entries.hasMoreElements()) {
                 entry = entries.nextElement();
                 String entryName = entry.getName();
-                File outFile = new File(zipDir + entryName);
+
                 // Verify that this file/directory will be extracted into our zipDir (and not somewhere else!)
-                if (!outFile.toPath().normalize().startsWith(zipDir)) {
+                Path validatedOutPath;
+                try {
+                    String fileAbsolutePath =
+                        SecureFileAccess.calculateAbsolutePathUsingBaseDir(entryName, zipDir);
+                    validatedOutPath = SecureFileAccess.validatePathForWrite(fileAbsolutePath, List.of(zipDir),
+                                                                             "ItemImport zip entry extraction");
+                } catch (IOException e) {
                     throw new IOException("Bad zip entry: '" + entryName
                                               + "' in file '" + zipfile.getAbsolutePath() + "'!"
-                                              + " Cannot process this file or directory.");
+                                              + " Cannot process this file or directory.", e);
                 }
+
+                File outFile = validatedOutPath.toFile();
+
                 if (entry.isDirectory()) {
                     if (!outFile.mkdirs()) {
                         logError("Unable to create contents directory: " + zipDir + entry.getName());
@@ -2048,6 +2063,13 @@ public String unzip(File zipfile, String destDir) throws IOException {
                     out.close();
                 }
             }
+        } catch (Exception e) {
+            // If an error occurs in unzipping, cleanup the directory we were unzipping this file into
+            if (sourcedirFile.exists()) {
+                FileUtils.deleteDirectory(sourcedirFile);
+            }
+            // Then throw error upwards
+            throw e;
         } finally {
             //Close zip file
             zf.close();

From c6451329c874236ea19a2a16a5785f06c67fab68 Mon Sep 17 00:00:00 2001
From: Tim Donohue <tim.donohue@lyrasis.org>
Date: Thu, 6 Aug 2026 15:45:29 -0500
Subject: [PATCH 3/4] Add a basic "fail fast" validation of file paths for ALL
 scripts. This may not offer full protection against path traversal, but it
 makes it more difficult by validating files passed into scripts cannot use
 path characters in their name.

(cherry picked from commit f005011a99a84871ba7a2777d8e035aa436cec65)
---
 .../org/dspace/app/rest/repository/ScriptRestRepository.java | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ScriptRestRepository.java b/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ScriptRestRepository.java
index 848ddfbac9b3..a97e9e2e3690 100644
--- a/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ScriptRestRepository.java
+++ b/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ScriptRestRepository.java
@@ -193,6 +193,11 @@ private void checkFileNames(DSpaceRunnable dSpaceRunnable, List<MultipartFile> f
             String fileName = file.getOriginalFilename();
             if (fileNames.contains(fileName)) {
                 throw new UnprocessableEntityException("There are two files with the same name: " + fileName);
+            } else if (StringUtils.containsAny(fileName, "..", "/", "\\")) {
+                // NOTE: This is just a "fail-fast", basic protection against a potential path traversal attack.
+                // Individual scripts MUST also contain their own path traversal protections.
+                throw new UnprocessableEntityException(
+                    "Filenames cannot contain path characters such as '..', '/', or '\\' : " + fileName);
             } else {
                 fileNames.add(fileName);
             }

From 9a531e27d17a3af1188a0cc8d3e13654c3586c04 Mon Sep 17 00:00:00 2001
From: Tim Donohue <tim.donohue@lyrasis.org>
Date: Thu, 6 Aug 2026 15:49:00 -0500
Subject: [PATCH 4/4] Add an integration test that proves traversal attacks are
 now blocked in ItemImport. This test fails on current "main" branch

(cherry picked from commit 1a2844bc909ea8a07fe34be308f3484062b31817)
---
 .../dspace/app/itemimport/ItemImportIT.java   | 53 +++++++++++++++++++
 1 file changed, 53 insertions(+)

diff --git a/dspace-server-webapp/src/test/java/org/dspace/app/itemimport/ItemImportIT.java b/dspace-server-webapp/src/test/java/org/dspace/app/itemimport/ItemImportIT.java
index 2cd93deafd3e..3de35a3b69b4 100644
--- a/dspace-server-webapp/src/test/java/org/dspace/app/itemimport/ItemImportIT.java
+++ b/dspace-server-webapp/src/test/java/org/dspace/app/itemimport/ItemImportIT.java
@@ -59,6 +59,7 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.MediaType;
 import org.springframework.mock.web.MockMultipartFile;
+import org.springframework.test.web.servlet.ResultMatcher;
 
 /**
  * Basic integration testing for the SAF Import feature via UI {@link ItemImport}.
@@ -137,6 +138,39 @@ public void importItemByZipSafWithBitstreams() throws Exception {
         assertTrue(workTempDir.exists());
     }
 
+    @Test
+    public void importItemUsingInvalidZipName() throws Exception {
+        String zipfileName = "saf-bitstreams.zip";
+        // This zip name is invalid as it must only contain a file name.
+        // This is an example path traversal attack.
+        String badZipFileName = "../../../" + zipfileName;
+
+        // First attack attempt. Provide a path traversal via both "-z" param and "filename" of multipart request.
+        LinkedList<DSpaceCommandLineParameter> parameters = new LinkedList<>();
+        parameters.add(new DSpaceCommandLineParameter("-a", ""));
+        parameters.add(new DSpaceCommandLineParameter("-c", collection.getID().toString()));
+        parameters.add(new DSpaceCommandLineParameter("-z", badZipFileName));
+        MockMultipartFile bitstreamFile = new MockMultipartFile("file", badZipFileName,
+                                                                MediaType.APPLICATION_OCTET_STREAM_VALUE,
+                                                                getClass().getResourceAsStream("saf-bitstreams.zip"));
+
+        // Verify this import will fail and throw a 422 error
+        performImportScriptExpectError(parameters, bitstreamFile, status().isUnprocessableEntity());
+
+        // Second attack attempt. Provide a path traversal via "-z" param but a VALID file in "filename"
+        // of multipart request.
+        parameters = new LinkedList<>();
+        parameters.add(new DSpaceCommandLineParameter("-a", ""));
+        parameters.add(new DSpaceCommandLineParameter("-c", collection.getID().toString()));
+        parameters.add(new DSpaceCommandLineParameter("-z", badZipFileName));
+        bitstreamFile = new MockMultipartFile("file", zipfileName,
+                                              MediaType.APPLICATION_OCTET_STREAM_VALUE,
+                                              getClass().getResourceAsStream("saf-bitstreams.zip"));
+
+        // Verify this import will also fail and throw a 422 error
+        performImportScriptExpectError(parameters, bitstreamFile, status().isUnprocessableEntity());
+    }
+
     @Test
     public void importItemByZipSafWithRelationships() throws Exception {
         context.turnOffAuthorisationSystem();
@@ -266,4 +300,23 @@ private void checkProcess(Process process) {
                 .filter(b -> StringUtils.contains(b.getName(), ".zip"))
                 .count());
     }
+
+    private void performImportScriptExpectError(LinkedList<DSpaceCommandLineParameter> parameters,
+                                               MockMultipartFile bitstreamFile,
+                                               ResultMatcher expectedErrorMatcher)
+        throws Exception {
+        List<ParameterValueRest> list = parameters.stream()
+                                                  .map(dSpaceCommandLineParameter -> dSpaceRunnableParameterConverter
+                                                      .convert(dSpaceCommandLineParameter, Projection.DEFAULT))
+                                                  .collect(Collectors.toList());
+
+        AtomicReference<Integer> idRef = new AtomicReference<>();
+
+        // Verify process fails
+        getClient(getAuthToken(admin.getEmail(), password))
+            .perform(multipart("/api/system/scripts/import/processes")
+                         .file(bitstreamFile)
+                         .param("properties", mapper.writeValueAsString(list)))
+            .andExpect(expectedErrorMatcher);
+    }
 }
